Firebase.cpp

```diff
@@ -15,100 +15,241 @@
+  String errorMessage() const;
+
+  // True if http status code is 200(OK).
+  bool isOk() const;
+
+  // Message sent back from Firebase backend. This pulls value to local memory,
+  // be careful if value can be large.
+  String rawResponse();
+
+  int httpStatus() const {
+    return status_;
+  }
+
+ private:
+  FirebaseCall(HTTPClient* http);
+
+  HTTPClient* http_;
+
+  int status_;
+  String error_message_;
+};
+
+}  // namespace
+
+Firebase::Firebase(const String& host) : host_(host) {
+  http_.setReuse(true);
 }
 
 Firebase& Firebase::auth(const String& auth) {
-  _auth = auth;
+  auth_ = auth;
   return *this;
 }
 
-String Firebase::get(const String& path) {
-  return sendRequest("GET", path);
+FirebaseGetResult Firebase::get(const String& path) {
+  FirebaseCall call(host_, auth_, "GET", path, &http_);
+  return call.isError() ? FirebaseGetResult::FromError(call.errorMessage())
+      : FirebaseGetResult::FromResponse(call.rawResponse());
 }
 
-String Firebase::push(const String& path, const String& value) {
-  return sendRequest("POST", path, value);
+FirebasePushResult Firebase::push(const String& path, const String& value) {
+  FirebaseCall call(host_, auth_, "POST", path, value, &http_);
+  return call.isError() ? FirebasePushResult::FromError(call.errorMessage())
+      : FirebasePushResult::FromResponse(call.rawResponse());
 }
 
-Firebase& Firebase::stream(const String& path) {
-  _error.reset();
-  String url = makeURL(path); 
-  const char* headers[] = {"Location"};
-  _http.setReuse(true);  
-  _http.begin(_host.c_str(), firebasePort, url.c_str(), true, firebaseFingerprint);
-  _http.collectHeaders(headers, 1);
-  _http.addHeader("Accept", "text/event-stream");
-  int statusCode = _http.sendRequest("GET", (uint8_t*)NULL, 0);
-  String location;
-  // TODO(proppy): Add a max redirect check
-  while (statusCode == 307) {
-    location = _http.header("Location");
-    _http.setReuse(false);
-    _http.end();
-    _http.setReuse(true);    
-    _http.begin(location, firebaseFingerprint);
-    statusCode = _http.sendRequest("GET", (uint8_t*)NULL, 0);
+FirebaseRemoveResult Firebase::remove(const String& path) {
+  FirebaseCall call(host_, auth_, "DELETE", path, &http_);
+  if (call.isError()) {
+    return FirebaseRemoveResult::FromError(call.errorMessage());
   }
-  if (statusCode != 200) {
-    _error.set(statusCode,
-	       "stream " + location + ": "
-	       + HTTPClient::errorToString(statusCode));
+  // Remove is only complete if returned status is OK(200).
+  if (!call.isOk()) {
+    return FirebaseRemoveResult::FromError(
+        "Remove " + path + " returned with status " + call.httpStatus());
   }
-  return *this;
+  return FirebaseRemoveResult::Ok();
 }
 
-String Firebase::makeURL(const String& path) {
-  String url;
-  if (path[0] != '/') {
-    url = "/";
-  }
-  url += path + ".json";
-  if (_auth.length() > 0) {
-    url += "?auth=" + _auth;
+FirebaseEventStream Firebase::stream(const String& path) {
```

Comment:
Well since it is an interactive object it seemed like less of a "result" but this may be to your point of not naming any of them "result" because they may be interactive I. the future with laziness.